design-documents/provisioning/configuration.md

```diff
@@ -0,0 +1,238 @@
+* Recurring Data
+* app:config:import
+
+`bin/magento setup:upgrade`
+
+* Declarative Schema (DB)
+* Install Schema
+* Upgrade Schema
+* Recurring Schema
+* Install Data
+* Upgrade Data
+* Recurring Data
+* app:config:import
+
+There is a dependency tree during installation:
+
+* Declarative Schema 
+* Schema Patches 
+* Recurring Schema Scripts
+* Websites => Store Groups => Stores 
+* System Configuration 
+* Attributes => Attribute Sets / Attribute Groups
+* Tax 
+* Data Patches
+* Recurring Data Scripts
+* app:config:import
+
+
+#### Data structure
+
+As already mentioned, the data structure is related to the existing database declarative schema. Every module can define a XML file for a particular configuration type. All configuration files for the same type would be merged to a single logical structure. Therefore every configuration type module should at least support merge logic. This is done individually, as for example, websites are merged differently than attributes.
+
+Therefore, a virtual type of `Magento\Framework\Config\Reader\Filesystem` will be defined, which has a ` Magento\Framework\Config\ConverterInterface` which is responsible for merging the XML files.
+
+#### Possibility of admin side changes
+
+The admin should still be allowed to change certain data from admin side. And some data shouldn't be changed by admin.
+
+Therefore there is a field `is_mutable` which defines if the declared data can be changed by the admin. There is a checkbox on admin side which can be unchecked to change the data. This is similar to the "use system value" in the store configuration. The following explains the behavior:
+
+|                         | `is_mutable = true`                 | `is_mutable = false` |
+| ------------------------| ----------------------------------- | -------------------- |
+| new entity is installed | yes                                 | yes                  |
+| checkbox on admin side  | exists and initially checked        | not visible          |
+| admin can change entity | yes                                 | no                   |
+| save btn                | disabled (until checkbox unchecked) | disabled             |
+| delete btn              | disabled (until checkbox unchecked) | disabled             |
+
+The `is_mutable` flag is always `false` in a default behavior, that would mean if it is not explicitly mentioned by XML, the admin can not change the entity.
+
+If XML with the same attribute code will be created (because it was created by admin first) it will take over all responsibilities to a code side and will be then managed like attribute created by XML declaration.
+
+**Optional behaviour:** If admin unchecked the checkbox and changes the entity, the data can be restored if the admin checks the checkbox again. The data from XML delcaration will then restore the entity. This behaviour is optional as it is not really necessary regarding a provisioning functionality, but of course nice to have.
```

Comment:
When the data is restored? At the moment the admin unchecks the checkbox or during next update? If at the moment of the admin action, won't be it too destructive to the store?